### DISCUSSION TITLE
Error of ~/anaconda3/envs/moose/libmesh/contrib/bin/libtool: 1: /usr/bin/sed: not found

### DISCUSSION THREAD

**Discussion by yw5aj** [idaholab/moose#16123] [Q&A Getting Started] [answered]
Dear all,

Sorry about this naive question...

Everything was working fine until earlier today that I updated MOOSE based on [this page](https://mooseframework.inl.gov/getting_started/installation/index.html#update). I made sure that `moose-libmesh` is up-to-date, re-pulled the github repo of `moose`, and then tried to run `make cloberrall`. 

But then I got the message of `/anaconda3/envs/moose/libmesh/contrib/bin/libtool: 1: /usr/bin/sed: not found`. Could anyone please help me out a bit?

Best,
Shawn

**Comment by jessecarterMOOSE** [comment]
Glad to hear I'm not the only one with this problem! @lindsayad @milljm did you ever figure out what was going on? I ended up just making a soft link.

**Reply by lindsayad** [reply]
No I don't have much of a clue on this one

**Reply by milljm** [reply]
I am not sure why libtool would hard code a path to sed like that. If it's Anaconda, that would be even stranger (implying that it would go about mucking with packages like this). I don't think it's libMesh's fault as well, as this is a 'contrib'. The other thing is, this wasn't an issue until recently. The Conda recipes creating moose-libmesh have not changed since it's inception.

**Comment by milljm** [comment]
@cticenhour reported this error happening on a few of our build machines (but then stopped). I suspect it occurred when that build machine found an issue with conda, and attempted to build the environment from scratch (such as what would happen with a new-user install).

I'll take some time today and try and reproduce the issue.

Edit: I can replicate this error... I'll look into it.

**Reply by cticenhour** [reply]
It's definitely still happening....I just haven't been able to get back to this (too many meetings!). Let me know how I can help (or provide another box to test something on)

**Reply by cticenhour** [reply]
Last data point I have was 10 hours ago on `mobuild31` for Freya.

**Comment by milljm** [comment]
It looks like libmesh/contrib/bin/libtool is generated during ./configure. And so is recording where it found `sed` at the time. Unfortunately, not all machines have `sed` in the same place. Not exactly sure why we didn't see this error immediately after the first version of moose-libmesh rolled out the door.

What I think we need to do, is include `sed` as a requirement when building/running libMesh. That way when moose-libmesh is installed, the dependency `sed` will also be installed, and hopefully be available in the same location as it was during build. Conda does some magical things when it prepares a package... hopefully it corrects more than just link lines (libtool is a script, and I am not sure Conda sniffs out paths inside scripts when it corrects/prepares them for redistribution)

**Comment by laumiulun** [comment]
softlink kind of bandage solve the problem, but it keep showing the message:

'libtool:   error: ignoring unknown tag CXX' 

It still finishes compiling and the test run fine.

**Comment by milljm** [comment]
When I build moose-libmesh and set a dependency on sed, it works. So I will commit a change to moose/conda/libmesh/meta.yaml, unless @lindsayad has a better idea,

**Reply by milljm** [reply]
scratch that. It works. But I am spammed by a new message:

```pre
libtool:   error: ignoring unknown tag CXX
```

Still investigating.

**Reply by lindsayad** [reply]
I would like to try using a conda/system provided libtool per the slack discussion, but that would require some changes in the MOOSE makefiles

**Reply by laumiulun** [reply]
I didn't encounter this issue if I started from scratch in a VM, only when I am updating moose & conda. Nuking the conda environment didn't help either.

**Comment by milljm** [comment]
A couple of work-arounds:

Downgrade to the previous version of moose-libmesh:

```bash
conda install moose-libmesh=2020.10.05
```

Or build your own libMesh:

```bash
conda deactivate
conda create -n mylibmesh moose-petsc moose-tools
conda activate mylibmesh # use this instead of 'moose' for now...
cd moose
export MOOSE_JOBS=8   # what ever core count you use
scripts/update_and_rebuild_libmesh.sh
```

**Reply by jessecarterMOOSE** [reply]
After downgrading libmesh, would one need to downgrade moose accordingly? If so what hash?

**Reply by milljm** [reply]
I can't really provide a specific hash. As it all depends on what branch you are operating on. What I can say is that if you restricted your checkouts to within the same time frame of 2020-10-05 through 2020-11-03 (the dates for which the previous moose-libmesh package was vetted for) you should be safe.

The ideal work-around is to build libMesh yourself however, thus always marching forward.

**Reply by jessecarterMOOSE** [reply]
Yeah I figured. I just wasn't sure if the newest moose works with an older libmesh.

**Reply by lindsayad** [reply]
Hard to say. Our MOOSE-libMesh integration is typically pretty tight. My only way of knowing would be to go try it. I'd put it at about 50% odds of being successful. Sometimes I'm aware of backwards incompatible changes in libMesh...and I think we had one of those not this libMesh update but the one prior. So you might be good to go...

**Reply by mcacace** [reply]
Hi,
FYI ...
Yesterday I found that manual installation (petsc and libmesh) worked in the end --> this was my "easy" work-around after a full half day of trying.
I'm currently testing the second work-around as per @milljm post (my bad not having read it before). One thing I noticed is that in order to compile libmesh (despite activating petsc and the moose utils from the env in conda), I must set all CPPFLAGS, CXXFLAGS, CFLAGS, FFLAGS. Before I was having errors related to the metaphysicl package. 
So far configuring libmesh seems to run (not yet finished though) and will see if moose also compiles fine.
Thanks again for the support,
Mauro

**Reply by mcacace** [reply]
quick update, it did not work. Will investigate it further

**Reply by mcacace** [reply]
solved the issue (conflicting boost libraries). Moose compiled fine by configuring libmesh only as well. 
Sorry for the multiple posts.

**Reply by milljm** [reply]
That was interesting that you had to set all those CFLAGS... Apologies about the delayed response. I think my GitHub account and email notifications are _now_ set correctly.

Can you post the error which occurred during your attempt to build libMesh?

**Comment by lindsayad** [comment]
Ok I've opened an issue for this at #16232

**Comment by milljm** [comment]
I built moose-libmesh manually and have made it available on Anaconda. Please remove moose-libmesh (if you're using it) and re-install the environment (after clearing your download cache to make sure you get the latest version from anaconda).

```bash
conda deactivate   # make sure you are not using moose
conda env remove -n moose --yes
conda clean --all --yes
conda create -n moose moose-libmesh moose-tools
```


Edit: When I originally posted this, it didn't work. I have since corrected that error (verified on Ubuntu 18)

**Reply by tfc1992** [reply]
> I built moose-libmesh manually and have made it available on Anaconda. Please remove moose-libmesh (if you're using it) and re-install the environment (after clearing your download cache to make sure you get the latest version from anaconda).
> 
> ```shell
> conda deactivate   # make sure you are not using moose
> conda env remove -n moose --yes
> conda clean --all --yes
> conda create -n moose moose-libmesh moose-tools
> ```

I tried these commands, but still failed to compile. what am I missing?

**Reply by GiudGiud** [reply]
Compiling moose I assume. Same error message as up top?

**Reply by tfc1992** [reply]
> Compiling moose I assume. Same error message as up top?
yes, the same error message.

**Reply by tfc1992** [reply]
Any suggestions to solve this issue?

**Reply by GiudGiud** [reply]
Go for the build your own libmesh solution for now. It wont stop you from going back the next time we update moose's version of libmesh

**Reply by tfc1992** [reply]
Ok, thanks a lot.

**Reply by milljm** [reply]
@tfc1992 What make and version of Linux are you using by chance?

**Reply by tfc1992** [reply]
@milljm  make v 4.1 and ubuntu 18.01

**Reply by milljm** [reply]
Thank you!

**Reply by milljm** [reply]
I am building another version of moose-libmesh at the moment. I am going to just hack my way through it as suggested by @dschwen. I'll post again when it's complete.

**Comment by tfc1992** [comment]
Many thanks. I am a new moose user.

**Reply by milljm** [reply]
okay, finished... if you would perform the following and try again:

```bash
conda update --all
```

This is assuming you have not already build your own version of libMesh as suggested by @GiudGiud.

**Reply by tfc1992** [reply]
Wow, great, the compilation was successful, thank you very much! @milljm

**Reply by milljm** [reply]
Welcome!

**Comment by milljm** [comment, marked as answer]
For those still receiving this error, please perform an update to get the latest moose-libmesh package:

```bash
conda update --all
```

**Comment by jessecarterMOOSE** [comment]
@milljm Does the new moose-libmesh package only contain a fix to the sed/grep issue or are there other changes I need?

**Reply by loganharbour** [reply]
To hop in - it also contains the changes in https://github.com/idaholab/moose/pull/16375 if you haven't updated in a while (beginning of Nov)

We've gotten bad about sending out emails for libmesh updates.

**Reply by jessecarterMOOSE** [reply]
Ah, I'll update it then.

Just curious, do you still plan on notifying people about libmesh updates and if so how if you're not using the google group?

**Reply by jessecarterMOOSE** [reply]
I just wonder how many people are "subscribed" to these github discussions.

**Reply by milljm** [reply]
That is a good question. The short answer is that right at this moment, it contains actual changes to libMesh (as Logan just mentioned). Though had you run `conda update --all` at 3AM on Tuesday when I created the fix manually... no it would not have. That package at 3AM only contained the fix for sed|grep|dd....

**Reply by jessecarterMOOSE** [reply]
> Though had you run `conda update --all` at 3AM on Tuesday when I created the fix manually...  

Nope definitely didn't do that. 

I had added some softlinks for grep/sed to fix the problem on my system and was just wondering if there was more I needed to do.
